File: day04/main.go
package main

import (
	"fmt"
)

const (
	puzzleInputMin = 231832
	puzzleInputMax = 767346
)

func passToDigits(pass int) [6]int {
	return [6]int{
		pass / 100000 % 10,
		pass / 10000 % 10,
		pass / 1000 % 10,
		pass / 100 % 10,
		pass / 10 % 10,
		pass % 10,
	}
}

func isValidPass(pass int) bool {
	s := passToDigits(pass)

	hasDouble := false

	k := s[0]
	for _, i := range s[1:] {
		if i == k {
			hasDouble = true
		}
		if i < k {
			return false
		}
		k = i
	}

	return hasDouble
}

func isValidPass2(pass int) bool {
	if !isValidPass(pass) {
		return false
	}

	s := passToDigits(pass)

	hasRun2 := false

	run := 1
	k := s[0]
	for _, i := range s[1:] {
		if i == k {
			run++
		} else {
			if run == 2 {
				hasRun2 = true
			}
			run = 1
		}
		k = i
	}

	return hasRun2 || run == 2
}

func main() {
	count := 0
	count2 := 0
	for i := puzzleInputMin; i <= puzzleInputMax; i++ {
		if isValidPass(i) {
			count++
		}
		if isValidPass2(i) {
			count2++
		}
	}
	fmt.Println(count)
	fmt.Println(count2)
}

File: day04/src/main.rs
const PUZZLE_INPUT_MIN: i32 = 231832;
const PUZZLE_INPUT_MAX: i32 = 767346;

fn pass_to_digits(pass: i32) -> [i32; 6] {
    [
        pass / 100000 % 10,
        pass / 10000 % 10,
        pass / 1000 % 10,
        pass / 100 % 10,
        pass / 10 % 10,
        pass % 10,
    ]
}

fn is_valid_pass(pass: i32) -> bool {
    let (_, double, decr) = pass_to_digits(pass)
        .into_iter()
        .fold((-1, false, false), |(prev, has_double, decr), &i| {
            (i, has_double || i == prev, decr || i < prev)
        });
    double && !decr
}

fn is_valid_pass2(pass: i32) -> bool {
    let (_, run, has_run2) =
        pass_to_digits(pass)
            .into_iter()
            .fold((-1, 0, false), |(prev, run, has_run2), &i| {
                if i == prev {
                    (i, run + 1, has_run2)
                } else {
                    (i, 1, has_run2 || run == 2)
                }
            });
    (has_run2 || run == 2) && is_valid_pass(pass)
}

fn main() {
    let k1 =
        (PUZZLE_INPUT_MIN..=PUZZLE_INPUT_MAX)
            .fold(0, |s, i| if is_valid_pass(i) { s + 1 } else { s });
    println!("{}", k1);
    let k2 =
        (PUZZLE_INPUT_MIN..=PUZZLE_INPUT_MAX)
            .fold(0, |s, i| if is_valid_pass2(i) { s + 1 } else { s });
    println!("{}", k2);
}
